source/_posts/classify-mnist-digits-keras.md

```diff
@@ -0,0 +1,334 @@
+
+The input then is passed to a **convolutional layer** using `n` filters, it creates another matrix with different height, width and as deep as many filters defined. A filter is a small window normally *5x5xinput_depth* in size, it traverses all the image to get its features. Its output goes into and activation function, in this case a <a href="https://en.wikipedia.org/wiki/Rectifier_(neural_networks)">ReLu function</a>.
+
+The resulting layers then goes into a **pooling layer** that reduces dimensionality. Then the **fully-connected layer** is a normal feed forward layer for classification.
+
+We could combine the conv layer with another conv to a pooling layer and repeat it *n* times before going to the fully-connected layer that at the same time could have several fully-connected layers.
+
+<img src="/classify-mnist-digits-keras/conv.png" />
+
+For further detail about Convolutiona Neural Networks you can visit http://cs231n.stanford.edu/ or go the [bibligraphy section](#bibliography).
+
+
+## Libraries and modules
+
+```python
+import numpy as np
+np.random.seed(123)  # for reproducibility
+
+from keras.models import Sequential
+from keras.layers import Dense, Dropout, Flatten
+from keras.layers import Conv2D, MaxPooling2D
+from keras.utils import np_utils
+from keras.datasets import mnist
+
+from matplotlib import pyplot as plt
+```
+
+I import [Numpy](http://www.numpy.org/) for matrix manipulation, [pyplot](https://matplotlib.org/api/pyplot_api.html) from [matplotlib](https://matplotlib.org/) to render some graphic representation of the data, `keras.util` to more data manipulation, and the [MNIST dataset](https://keras.io/datasets/#mnist-database-of-handwritten-digits) provided by keras.
+
+The keras [Sequential](https://keras.io/models/sequential/) model allows to `add` layers, each layer having its own arquitecture and purpose.
+
+The idea is to sequentially add different layers to our model util having a Neural Network arquitecture that can process and classify images of handwritting digits.
+
+## Data Preparation
+
+Every Machine Learning model needs some data preparation. Sometimes this is the most exahustive and underrated (for beginners) part of the process.
+
+1. Load dataset.
+2. Separate dataset into training and test datasets.
+3. Visualize data to get some intuition.
+4. Prepare input data to feed the input layer.
+5. Prepare label data.
+
+```python
+# load dataset
+(X_train, y_train), (X_test, y_test) = mnist.load_data();
+```
+
+The `load_data()` function return two tuples, already divinding the data in training and test collections.
+
+The `X_train.shape` will be `(60000, 28, 28)`, 60,000 examples of 28x28 pixes grayscale images of the 10 digits. While `y_train` is an vector fo 60,000 examples where each element is an integer from 0 to 9 (`X_test` and `y_test` contains 10,000 examples). Remember that `X_train[n]` corresponds to `y_train[n]`.
```

Comment:
One tip on parentheses, nothing to change: http://thecopyfox.com/2012/11/writing-tip-do-you-overuse-parentheses/